### PR TITLE
Create ultralytics_yolov5.md

### PR DESCRIPTION
**Pull request by glenn-jocher**
🚀 This PR submits Ultralytics YOLOv5 for consideration for inclusion into PyTorch Hub. 

We've verified correct operations of the tutorials presented in this PR, and also added a PyTorch Hub section directly to our README at https://github.com/ultralytics/yolov5#pytorch-hub to improve awareness understanding of this feature.

We've streamlined YOLOv5 workflows to allow for an absolute minimum viable code to run full inference on PIL or OpenCV images (including batched inference) with automatic source handling, image reshaping/padding, NMS and post-processing all included and handled automatically:

```python
import torch
from PIL import Image

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True).fuse().eval()  # yolov5s.pt
model = model.autoshape()  # for autoshaping of PIL/cv2/np inputs and NMS

# Images
torch.hub.download_url_to_file('https://github.com/ultralytics/yolov5/releases/download/v1.0/zidane.jpg', 'zidane.jpg')
img = Image.open('zidane.jpg')  # PIL image

# Inference
prediction = model(img)  # includes NMS
```

Let us know if any changes are required please. All feedback is welcome!

### CODE REVIEW

**Comment by glenn-jocher** [comment]
circleci checks are failing due to an unrelated problem that seems to originate with huggingface. I see an open PR on the topic as well https://github.com/pytorch/hub/pull/161. Netlify preview looks good.

```bash
Running pytorch example in huggingface_pytorch-transformers.md
Downloading: "https://github.com/huggingface/pytorch-transformers/archive/master.zip" to /home/circleci/.cache/torch/hub/master.zip
Traceback (most recent call last):
  File "temp.py", line 2, in <module>
    tokenizer = torch.hub.load('huggingface/pytorch-transformers', 'tokenizer', 'bert-base-cased')
  File "/home/circleci/miniconda3/lib/python3.7/site-packages/torch/hub.py", line 349, in load
    hub_module = import_module(MODULE_HUBCONF, repo_dir + '/' + MODULE_HUBCONF)
  File "/home/circleci/miniconda3/lib/python3.7/site-packages/torch/hub.py", line 71, in import_module
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/hubconf.py", line 8, in <module>
    from transformers import (
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/__init__.py", line 135, in <module>
    from .pipelines import (
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/pipelines.py", line 38, in <module>
    from .tokenization_auto import AutoTokenizer
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_auto.py", line 118, in <module>
    from .tokenization_albert_fast import AlbertTokenizerFast
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_albert_fast.py", line 23, in <module>
    from .tokenization_utils_fast import PreTrainedTokenizerFast
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/tokenization_utils_fast.py", line 30, in <module>
    from .convert_slow_tokenizer import convert_slow_tokenizer
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/convert_slow_tokenizer.py", line 27, in <module>
    from transformers.utils import sentencepiece_model_pb2 as model
  File "/home/circleci/.cache/torch/hub/huggingface_pytorch-transformers_master/src/transformers/utils/sentencepiece_model_pb2.py", line 9, in <module>
    from google.protobuf import descriptor as _descriptor
ModuleNotFoundError: No module named 'google'
```

**Comment by ailzhang** [comment]
@glenn-jocher Thanks for contributing to hub! I've fixed the dependency issue in a separate PR, would you mind rebasing on top of master and let the CI run again? Thanks!

**Comment by glenn-jocher** [comment]
@ailzhang thanks! I've rebased and committed bug fixes. CI checks are all green now. Netlify preview looks good, though I noticed the preview is missing the github star count.

**Comment by ailzhang** [comment]
gentlely ping @soumith on reviewing ;)

**Comment by glenn-jocher** [comment]
CI checks are failing on missing CLA and an apparent unrelated circleci check. Corporate CLA is now signed for Ultralytics LLC, with GitHub username glenn-jocher added.

File: ultralytics_yolov5.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+author: Ultralytics LLC
+tags: [vision, scriptable]
+github-link: https://github.com/ultralytics/yolov5
+github-id: ultralytics/yolov5
+featured_image_1: ultralytics_yolov5_img1.jpg
+featured_image_2: ultralytics_yolov5_img2.png
+accelerator: cuda-optional
+---
+
+🚀  This guide explains how to load YOLOv5 from PyTorch Hub [https://pytorch.org/hub/](https://pytorch.org/hub/)
```
Comment (soumith) [inline ultralytics_yolov5.md:18]: remove this line. seems redundant

File: ultralytics_yolov5.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+## Before You Start
+
+Start from a working python environment with **Python>=3.8** and **PyTorch>=1.6** installed, as well as `pyyaml>=5.3` for reading YOLOv5 configuration files. To install PyTorch see [https://pytorch.org/get-started/locally/](https://pytorch.org/get-started/locally/). To install dependencies:
+```bash
+$ pip install -U opencv-python pillow pyyaml tqdm  # install dependencies
+```
+
+Note that cloning the [https://github.com/ultralytics/yolov5](https://github.com/ultralytics/yolov5) repository is **not** required.
```
Comment (soumith) [inline ultralytics_yolov5.md:27]: remove this line. it's redundant for pytorch hub models

File: ultralytics_yolov5.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            str += '%g %ss, ' % (n, model.names[int(c)])  # add to string
+        for *box, conf, cls in pred:  # xyxy, confidence, class
+            label = model.names[int(cls)] if hasattr(model, 'names') else 'class_%g' % cls
+            # str += '%s %.2f, ' % (label, conf)  # label
+            ImageDraw.Draw(img).rectangle(box, width=3)  # plot
+    img.save('results%g.jpg' % i)  # save
+    print(str + 'Done.')
+```
+
+<img src="https://user-images.githubusercontent.com/26833433/96368758-114c3a00-1156-11eb-9ede-a03f9aff42e4.jpg" width="550">
```
Comment (soumith) [inline ultralytics_yolov5.md:92]: remove these two inlined images. they are not needed, as when the user opens the hub page in a notebook and runs them, they will see the output

File: ultralytics_yolov5.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+# Model
+model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True).fuse().eval()  # yolov5s.pt
+model = model.autoshape()  # for autoshaping of PIL/cv2/np inputs and NMS
+
+# Images
+torch.hub.download_url_to_file('https://github.com/ultralytics/yolov5/releases/download/v1.0/zidane.jpg', 'zidane.jpg')
+img = Image.open('zidane.jpg')  # PIL image
+
+# Inference
+prediction = model(img)  # includes NMS
```
Comment (soumith) [inline ultralytics_yolov5.md:48]: describe the output `prediction` here in a sentence.

File: ultralytics_yolov5.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+img1 = Image.open('zidane.jpg')  # PIL image
+img2 = cv2.imread('bus.jpg')[:, :, ::-1]  # OpenCV image (BGR to RGB)
+img3 = np.zeros((640, 1280, 3))  # numpy array
+imgs = [img1, img2, img3]  # batched list of images
+
+# Inference
+with torch.no_grad():
+    prediction = model(imgs, size=640)  # includes NMS
+
+# Plot
```
Comment (soumith) [inline ultralytics_yolov5.md:76]: break out the plotting into a separate code block

File: ultralytics_yolov5.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            ImageDraw.Draw(img).rectangle(box, width=3)  # plot
+    img.save('results%g.jpg' % i)  # save
+    print(str + 'Done.')
+```
+
+<img src="https://user-images.githubusercontent.com/26833433/96368758-114c3a00-1156-11eb-9ede-a03f9aff42e4.jpg" width="550">
+
+<img src="https://user-images.githubusercontent.com/26833433/97121810-54dc1080-1721-11eb-8c25-451ccb1452e4.jpg" width="350">
+
+For all inference options see YOLOv5 [`autoShape()` forward method](https://github.com/ultralytics/yolov5/blob/453acdec67202e15b772370d31eb196d844f2892/models/common.py#L129-L136):
```
Comment (soumith) [inline ultralytics_yolov5.md:96]: The PyTorch Hub card is not meant to be a complete reference. It is meant to be a card for gentle introduction.

The card has already gotten too long.
I suggest that from line 96 onwards, you remove everything, and instead add a pointer to your repository or documentation that showcases more options.

File: ultralytics_yolov5.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+```bash
+$ pip install -U opencv-python pillow pyyaml tqdm  # install dependencies
+```
+
+Note that cloning the [https://github.com/ultralytics/yolov5](https://github.com/ultralytics/yolov5) repository is **not** required.
+
+
+## Load YOLOv5 From PyTorch Hub
+
+### Simple Example
```
Comment (soumith) [inline ultralytics_yolov5.md:32]: let's aim to have just one example for the whole card, a simple example showcasing plotting as well.

So, please merge the Simple Example and Detailed Example sections into a single concise section.

File: ultralytics_yolov5.md [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True, classes=10)
+```
+In this case the model will be composed of pretrained weights **except for** the output layers, which are no longer the same shape as the pretrained output layers. The output layers will remain initialized by random weights.
+
+
+## Model Description
+
+<img width="800" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/97808084-edfcb100-1c64-11eb-83eb-ffed43a0859f.png">
+
+YOLOv5 is an architecture for compound-scaled object detection models trained on the COCO 2017 dataset, and includes built-in functionality for Test Time Augmentation (TTA), Model Ensembling, Rectangular Inference, Hyperparameter Evolution and export to a variety of common formats such as ONNX, CoreML and TFLite. YOLOv5 also features a free **iOS App** that demos real-time 30 FPS inference of all models on iPhones and iPads: [iDetection iOS App](https://apps.apple.com/us/app/idetection/id1452689527#?platform=iphone).
```
Comment (soumith) [inline ultralytics_yolov5.md:125]: you can summarize this description in a shorter form right before the `Load YOLOv5 From PyTorch Hub` section

File: ultralytics_yolov5.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
+
+- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
+- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/ultralytics/yolov5](https://www.kaggle.com/ultralytics/yolov5)
+- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart) 
+- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)
+
+
+## Citation
```
Comment (soumith) [inline ultralytics_yolov5.md:171]: keep the Citation and Contact sections

**Review by soumith — CHANGES_REQUESTED** [review]
Thanks for the contribution. It's really high quality and great.

The PyTorch Hub cards are meant to be short and concise and a gentle introduction.
With that in mind, I've requested some changes, with detailed comments in-line.

File: ultralytics_yolov5.md (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Done.

File: ultralytics_yolov5.md (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Done.

File: ultralytics_yolov5.md (hunk 3 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Done.

File: ultralytics_yolov5.md (hunk 5 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md:79]: Done.

File: ultralytics_yolov5.md (hunk 6 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Understood, update done.

File: ultralytics_yolov5.md (hunk 7 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Good point, update done.

File: ultralytics_yolov5.md (hunk 8 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md]: Yes, it's been shortened now.

File: ultralytics_yolov5.md (hunk 9 above)
Comment (glenn-jocher) [reply ultralytics_yolov5.md:95]: Ok!

**Comment by glenn-jocher** [comment]
@ailzhang @soumith thanks for the detailed review! I've gone ahead and implemented all requested changes.

**I've converted the PR into a draft temporarily** though as we are updating our torch hub functionality and want to include these updates in this PR. The updates will allow for much simplified plotting and printing of inference results, and will allow us to further simplify the card example. I think this may be ready in about a week's time. I will update here when that is complete.

**Comment by soumith** [comment]
sounds good

**Comment by glenn-jocher** [comment]
@soumith the YOLOv5 PyTorch Hub updates are complete now. We've updated our inference in https://github.com/ultralytics/yolov5/pull/1415 to make everything super easy, no more messy plotting code:

```
# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True).autoshape()  # for PIL/cv2/np inputs and NMS

# Images
img = Image.open('zidane.jpg')  # PIL image

# Inference
results = model(img)  # includes NMS

# View
results.show()  # .show() results, .save() jpgs, or .print() to screen
```

CI appears to be failing due to an unrelated huggingface_pytorch-transformers.md issue.

**Review by soumith — APPROVED** [review]
looks great! Thanks a lot for the contribution to PyTorch Hub.

**Comment by soumith** [comment]
the PR is still marked as Draft, please fix that to a final PR so that I can merge it.

Once I merge the PR, it will be live on the PyTorch website after 24 hours.

**Comment by glenn-jocher** [comment]
@soumith awesome, thank you!! :)

**Pull request merged**